### DISCUSSION TITLE
Use jax.debug.callback for profiling?

### DISCUSSION THREAD

**Discussion by mariogeiger** [jax-ml/jax#12262] [Q&A] [answered]
Hi, 

Thanks for the new profiling tools `jax.debug`! 👍 

I was wondering if I could use something like

```python
def print_time(message):
    print(f"\"{message}\" at {time.perf_counter():.3f} seconds")

@jax.jit
def f(x):
    jax.debug.callback(partial(print_time, "a"))
    x = jnp.sin(x)
    jax.debug.callback(partial(print_time, "b"))
    x = jnp.cos(x)
    jax.debug.callback(partial(print_time, "c"))
    return x
```

to profile my code but it happens that all the python functions seems to be called at the same time and therefore I cannot measure execution time of different part of my code.

Any suggestions?
Is it hopeless?

**Comment by mariogeiger** [comment, marked as answer]
I found a solution to my problem :)

```python
def print_time(message, x):
    print(f'"{message}" at {time.perf_counter():.3f} seconds')
    return x


@partial(jax.custom_jvp, nondiff_argnums=(0,))
def profile(message, x):
    return jax.pure_callback(partial(print_time, message), x, x)


@profile.defjvp
def profile_jvp(message, primals, tangents):
    (x,) = primals
    (dx,) = tangents
    return profile(message + ":x", x), profile(message + ":dx", dx)


@jax.jit
def f(x):
    x = profile("A", x)
    x = jnp.sin(x)
    x = profile("B", x)
    x = jnp.cos(x)
    x = profile("C", x)
    return x
```

**Comment by sharadmv** [comment]
In general `jax.debug.callback` should not be used for profiling because the compiler is free to reorder the callbacks and they may not happen in program order.

Although you used `jax.pure_callback` to sequentialize the callbacks via data-dependency, your profile timings will now include device-to-host copy times needed for the pure callback. 

For more general purpose profiling, I recommend taking a look at the [Profiling JAX Programs](https://jax.readthedocs.io/en/latest/profiling.html) guide.

**Reply by mariogeiger** [reply]
Thanks for the comment, good to know about the device-to-host copy.
About the [Profiling JAX Programs](https://jax.readthedocs.io/en/latest/profiling.html) guide, as far as I understood, the technique proposed is opaque to jit compiled functions.

**Reply by sharadmv** [reply]
It should work with JIT-compiled functions.
